planner/core/expression_rewriter.go

```diff
@@ -370,19 +370,32 @@ func (er *expressionRewriter) handleCompareSubquery(v *ast.CompareSubqueryExpr)
 	switch v.Op {
 	// Only EQ, NE and NullEQ can be composed with and.
 	case opcode.EQ, opcode.NE, opcode.NullEQ:
-		condition, er.err = er.constructBinaryOpFunction(lexpr, rexpr, ast.EQ)
-		if er.err != nil {
-			er.err = errors.Trace(er.err)
-			return v, true
-		}
 		if v.Op == opcode.EQ {
 			if v.All {
 				er.handleEQAll(lexpr, rexpr, np)
 			} else {
+				// a = any(subq) will be rewriten as a in (subq).
+				if rCol, ok := rexpr.(*expression.Column); ok && er.asScalar {
+					rCol.InOperand = true
```

Comment:
Should we also make `ScalarFunction` to have this field? In `EvalBool()`, we only check `Column`, but if the two columns are of different types, the constructed binary operator function may takes two cast function as its parameters.

```sql
TiDB(root@127.0.0.1:test) > create table t1(a bigint, b bigint);
Query OK, 0 rows affected (0.01 sec)

TiDB(root@127.0.0.1:test) > create table t2(a varchar(20), b varchar(20));
Query OK, 0 rows affected (0.01 sec)

TiDB(root@127.0.0.1:test) > desc select * from t1 where t1.a not in (select t2.a from t2);
+----------------------------+----------+------+-----------------------------------------------------------------------------------+
| id                         | count    | task | operator info                                                                     |
+----------------------------+----------+------+-----------------------------------------------------------------------------------+
| Projection_9               | 8000.00  | root | test.t1.a, test.t1.b                                                              |
| └─HashLeftJoin_10          | 8000.00  | root | anti semi join, inner:Projection_14, equal:[eq(cast(test.t1.a), cast(test.t2.a))] |
|   ├─Projection_11          | 10000.00 | root | test.t1.a, test.t1.b, cast(test.t1.a)                                             |
|   │ └─TableReader_13       | 10000.00 | root | data:TableScan_12                                                                 |
|   │   └─TableScan_12       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                       |
|   └─Projection_14          | 10000.00 | root | test.t2.a, cast(test.t2.a)                                                        |
|     └─TableReader_16       | 10000.00 | root | data:TableScan_15                                                                 |
|       └─TableScan_15       | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                       |
+----------------------------+----------+------+-----------------------------------------------------------------------------------+
8 rows in set (0.00 sec)
```

Reply:
Patch has been updated to handle scalar function as argument of `in` expression, PTAL